ftl/common/cache_runner.py

```diff
@@ -0,0 +1,232 @@
+
+        _cache = LANGUAGE_CACHES[language]
+        self._cache_name = constants.GLOBAL_CACHE_REGISTRY + '/' + _cache
+
+        self._reg = docker_name.Registry('gcr.io', strict=False)
+        self._creds = docker_creds.DefaultKeychain.Resolve(self._reg)
+        self._transport = transport_pool.Http(httplib2.Http, size=2)
+        self._cache = docker_name.Tag(self._cache_name, strict=False)
+
+        # retrieve mappings when initializing runner
+        self._mappings = self.read_mappings()
+        logging.info(self._mappings)
+
+    def _tag(self, tag):
+        return docker_name.Tag(self._cache_name + ':' + tag)
+
+    def populate_cache(self):
+        existing_entries = self.retrieve_cache_entries()
+        logging.info(existing_entries)
+
+        # Determine which images exist in the cache that should not be there,
+        # and remove them
+        self.remove_old_entries(existing_entries)
+
+        # Populate cache with new entries
+        self.populate_cache_entries(existing_entries)
+
+        # Finally, write back the new key mapping to the filesystem
+        # to be copied to GCS
+        self.write_mapping_to_workspace()
+
+    def read_mappings(self):
+        """read cache_key -> package tuple mappings from GCS config file
+        return map of key to package,which we'll use as lookup
+        when pushing images"""
+
+        r = requests.get('https://www.googleapis.com/storage/v1'
+                         '/b/{bucket}/o/{file}?alt=media'.format(
+                             bucket=MAPPING_BUCKET,
+                             file=MAPPING_FILE.format(language=self._language)
+                         ))
+
+        if not r.ok:
+            logging.error('Error retrieving mapping: %s' % r.text)
+        try:
+            return json.loads(r.text)
+        except ValueError:
+            # no mapping found: create new one
+            return {}
+
+    def retrieve_cache_entries(self):
+        # returns all images stored in the cache currently
+        with docker_image.FromRegistry(self._cache,
+                                       self._creds,
+                                       self._transport) as session:
+            return set(tag.rstrip() for tag in session.tags() if tag)
+
+    def remove_old_entries(self, existing_entries):
+        # for each existing entry in the mapping,
+        # if it isn't in the package list, remove it
+        for entry in existing_entries:
+            entry_info = self._mappings.get(entry, '')
+            if entry_info and entry_info not in self._packages:
+                logging.info('removing entry {0} from cache'.format(
+                    entry_info))
+                self._remove_entry(entry)
+
+    def _remove_entry(self, entry):
+        # delete entry from mapping and cache
+        del self._mappings[entry]
+        docker_session.Delete(self._tag(entry), self._creds, self._transport)
+
+    def populate_cache_entries(self, existing_entries):
```

Comment:
Should this check self._mapping instead of the existing_entries list passed in here?

Reply:
the only reason we check the existing_entries list here is to avoid rebuilding an image. at this point we've already evicted entries that shouldn't be there and removed them from the mapping